### PR TITLE
subsys: usb: add loopback function

### PR DESCRIPTION
**Pull request by jfischer-no**
Add loopback function. This function can be used to test
USB device drivers and device stack connected to linux host
and has the similar interface as "Gadget Zero" [1] of the Linux
kernel.

Use modprobe usbtest to load the module, see also [2] for the
description of the tests and for Vendor and Product ID of the
"Gadget Zero". The userspace tool testusb [3] is needed to start
the tests.

[1] linux/drivers/usb/gadget/function/f_loopback.c
[2] linux/drivers/usb/misc/usbtest.c
[3] linux/tools/usb/testusb.c

### CODE REVIEW

**Review by loicpoulain — COMMENTED** [review]
Note that according to USB specification, a function cannot issue a STALL packet in response to a a SETUP packet. A function must ACK the SETUP or ignore it (cf Control Transfers chapter).

**Comment by jfischer-phytec-iot** [comment]
@loicpoulain In what context do you mean it?

**Comment by loicpoulain** [comment]
https://github.com/zephyrproject-rtos/zephyr/pull/7435/commits/bcd517953919e34838482fd964d920e572ec9d5d

**Comment by jfischer-phytec-iot** [comment]
@loicpoulain This protocol STALL is valid, e.g. if the callback and buffer for vendor request are not proper initialized, then the stack should return STALL (in response to any IN or OUT transaction, see 8.5.3.4) because the device is unable to complete the command.

**Comment by loicpoulain** [comment]
Yes you're right, this looks good indeed because the STALL will not be sent in response to the SETUP transaction/stage but to the following one(s) (IN or OUT)...

**Comment by carlescufi** [comment]
CC @pawelzadrozniak you can use this to test your driver

File: subsys/usb/class/loopback.c [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+static u8_t loopback_buf[1024];
+
+static void loopback_bulk_out_cb(u8_t ep,
+				 enum usb_dc_ep_cb_status_code ep_status)
+{
+	u32_t bytes_to_read;
+
+	usb_read(ep, NULL, 0, &bytes_to_read);
+	SYS_LOG_DBG("ep 0x%x, bytes to read %d ", ep, bytes_to_read);
+	usb_read(ep, loopback_buf, bytes_to_read, NULL);
```
Comment (finikorg) [inline subsys/usb/class/loopback.c]: should it send data back to IN?

**Review by finikorg — CHANGES_REQUESTED** [review]
I would add parts from commit message to documentation (in a follow up commit). We should also mention how to attach usbtest driver to this USB device.
BTW: Are you able to pass tests? For quark_se most of tests are failing.

File: subsys/usb/class/loopback.c (hunk 1 above)
Comment (jfischer-phytec-iot) [reply subsys/usb/class/loopback.c]: IN endpoint (buffer) is always prepared for the IN token and independent of OUT endpoint.

File: subsys/usb/usb_descriptor.c [hunk 2]
```diff
@@ -760,6 +767,46 @@ static struct dev_common_descriptor common_desc = {
...
 #endif /* CONFIG_USB_DFU_CLASS */
+#ifdef CONFIG_USB_DEVICE_LOOPBACK
+	.loopback_cfg = {
+		/* Interface descriptor 0 */
+		.if0 = {
+			.bLength = sizeof(struct usb_if_descriptor),
+			.bDescriptorType = USB_INTERFACE_DESC,
+			.bInterfaceNumber = FIRST_IFACE_LOOPBACK,
+			.bAlternateSetting = 0,
+			.bNumEndpoints = 2,
```
[the remark this answers is not in the document]
Comment (jfischer-phytec-iot) [reply subsys/usb/usb_descriptor.c]: there are always two, directly below :-)

[the remark this answers is not in the document]
Comment (finikorg) [reply subsys/usb/usb_descriptor.c]: The point was to omit "magic numbers" when possible.

**Review by finikorg — APPROVED** [review]

**Comment by finikorg** [comment]
@jfischer-phytec-iot Thanks for docs, my mistake was that I changed vendor/product pair in menuconfig (and with wrong device requiring to change alt setting, etc). It would be good to add this info to Zephyr docs or Readme.

**Comment by carlescufi** [comment]
@jfischer-phytec-iot can you please rebase this? Aside from that it's ready to merge

**Comment by galak** [comment]
Should I merge this?

**Comment by carlescufi** [comment]
@galak yes please, this one flew under my radar

**Pull request merged**